// vite-plugin-html-inject/dist/index.cjs
'use strict';

const vite = require('vite');
const path = require('path');
const fs = require('fs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const path__default = /*#__PURE__*/_interopDefaultLegacy(path);
const fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);

const tagMatcher = new RegExp('<load(?:.*?)="([^"]+)"(.*?)/>', "gs");
const attrMatcher = new RegExp('(?:(?:\\s)?([a-z0-9_-]+)(?:="([^"]*)"|))', "gi");
const replaceAttrMatcher = new RegExp("{=[$]([a-z0-9_-]+)}", "gi");
function escapeRegExp(input) {
  return input.replace(/[.*+?^${}()|[\]\\]/g, "$&");
}
function injectHTML(cfg) {
  let config;
  const fileList = /* @__PURE__ */ new Set();
  async function renderSnippets(code, codePath) {
    if (!config) {
      return code;
    }
    const matches = code.matchAll(tagMatcher);
    for (const match of matches) {
      let [tag, url, attrs] = match;
      let root = config.root;
      if (url.startsWith(".")) {
        root = path__default.dirname(root + codePath);
      } else {
        url = "/" + url;
      }
      if (!(url.endsWith(".htm") || url.endsWith(".html"))) {
        ["html", "htm"].some((item) => {
          const fileName = "/index." + item;
          const filePath2 = vite.normalizePath(path__default.join(root, url, fileName));
          if (fs__default.existsSync(filePath2)) {
            url += fileName;
            return true;
          }
        });
      }
      const filePath = vite.normalizePath(path__default.join(root, url));
      if (cfg?.debug?.logPath) {
        console.log("Trying to include ", filePath);
      }
      fileList.add(filePath);
      let out = tag;
      try {
        let data = fs__default.readFileSync(filePath, "utf8");
        for (const attr of attrs.matchAll(attrMatcher)) {
          const attrRegExp = new RegExp("{=\\$" + escapeRegExp(attr[1]) + "}", "g");
          data = data.replace(attrRegExp, attr[2]);
        }
        data = data.replace(replaceAttrMatcher, cfg?.replace?.undefined ?? "$&");
        out = await renderSnippets(data, url);
      } catch (error) {
        if (error instanceof Error) {
          out = error.message;
        }
        console.error(out);
      }
      code = code.replace(tag, out);
    }
    return code;
  }
  return {
    name: "static-html-loader",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },
    handleHotUpdate({ file, server }) {
      if (fileList.has(file)) {
        server.ws.send({
          type: "full-reload",
          path: "*"
        });
      }
    },
    transformIndexHtml: {
      enforce: "pre",
      transform(html, ctx) {
        return renderSnippets(html, ctx.path);
      }
    }
  };
}

module.exports = injectHTML;
